system/Common.php

```diff
@@ -289,7 +289,7 @@ function csrf_hash(): string
      */
     function csrf_field(?string $id = null): string
     {
-        return '<input type="hidden"' . ($id !== null ? ' id="' . esc($id, 'attr') . '"' : '') . ' name="' . csrf_token() . '" value="' . csrf_hash() . '"' . _solidus() . '>';
+        return '<input type="hidden"' . ((string) $id !== '' ? ' id="' . esc($id, 'attr') . '"' : '') . ' name="' . csrf_token() . '" value="' . csrf_hash() . '"' . _solidus() . '>';
```

Comment:
Why are we changing this?

Reply:
Because an empty `"" $id` doesn't make sense. We skip processing with the `esc()` function.

Reply:
How often we will set an empty string as a parameter here? Probably never.

It's already not very readable and with added typing it becomes even less. The world will not end if we escape the empty string.

I admit that this may just be my preference, but using typing everywhere makes the code less clear.

Reply:
That's why I've spoken out about rejecting `null `properties. The check would just be like` !== ""`
By the way, I noticed for the sake of interest the speed of the check - calling functions for an empty string imposes double the cost 0.02 vs 0.047 Yes, it's a small thing, just so you know

Reply:
In this case, you can skip the `instanceof `checks in many places, because checking for `null `is enough. Right? However, there are strict **phpspan** and **rector** rules to ensure that you can work with a valid object. I wanted to introduce stricter typing. Since null seems to be an "archaism from CI3 and PHP < 8"

Reply:
The "problem" is that an empty string is an edge case, not something we will experience regularly.

But casting to string will be made much more often + readability issues.

> In this case, you can skip the instanceof checks in many places, because checking for null is enough. Right? However, there are strict phpspan and rector rules to ensure that you can work with a valid object. I wanted to introduce stricter typing. Since null seems to be an "archaism from CI3 and PHP < 8"

I don't get what you're referring to here. And no, `null` is not archaism from CI3.

Does PHPStan or Rector have anything against checking that the value is not `null`? We are talking about specific cases.

Reply:
Is it possible to allow breaking changes? `csrf_field(string $id = '')`

Reply:
Nope. It will not happen in the `develop` branch. AFAIR we already changed the code once to include `null` values, but I would have to look for reasons behind this.

If PHPStan does not complain about it, it should not be part of this PR.